dns_sd_bonjour.c

```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
 		goto stop_browsing;
 	}
-	
-	if (dd->have_v4 || dd->have_v6) {
-		// Already resolved one IIO service. Stop.
-		DEBUG("DNS SD: first service already resolved. Skip.\n");
-		goto stop_browsing;
+
+	if ((flags & kCFNetServiceFlagRemove) != 0) {
+		DEBUG("DNS SD: Callback to remove service. Ignore.\n");
+		return;
 	}
 
+	iio_mutex_lock(dd->lock);
+
 	const CFNetServiceRef netService = (CFNetServiceRef)domainOrService;
 	if (netService == NULL) {
-		ERROR("DNS SD: Net service is null.\n");
-		goto stop_browsing;
+		DEBUG("DNS SD: Net service is null.\n");
+		goto verify_flags;
 	}
 	
 	if (!CFNetServiceResolveWithTimeout(netService, 10.0, &anError)) {
-		ERROR("DNS SD: Resolve error: %ld.%d\n", anError.domain, anError.error);
-		goto stop_browsing;
+		DEBUG("DNS SD: Resolve error: %ld.%d\n", anError.domain, anError.error);
+		goto exit;
 	}
 	
-	dd->port = CFNetServiceGetPortNumber(netService);
+	CFStringRef targetHost = CFNetServiceGetTargetHost(netService);
+	if (targetHost == NULL) {
+		DEBUG("DNS SD: No valid target host for service.\n");
+		goto exit;
+	}
+
+	char hostname[MAXHOSTNAMELEN];
+	if (!CFStringGetCString(targetHost, hostname, sizeof(hostname), kCFStringEncodingASCII)) {
+		ERROR("DNS SD: Could not translate hostname\n");
+		goto exit;
+	}
+
+	CFStringRef svcName = CFNetServiceGetName(netService);
+	char name[MAXHOSTNAMELEN];
+	if (!CFStringGetCString(svcName, name, sizeof(name), kCFStringEncodingASCII)) {
+		ERROR("DNS SD: Could not translate service name\n");
+		goto exit;
+	}
+
+	SInt32 port = CFNetServiceGetPortNumber(netService);
 	
 	CFArrayRef addrArr = CFNetServiceGetAddressing(netService);
 	if (addrArr == NULL) {
-		ERROR("DNS SD: No valid addresses for service.\n");
-		goto stop_browsing;
+		WARNING("DNS SD: No valid addresses for service %s.\n", name);
+		goto exit;
 	}
 	
+	bool	have_v4 = FALSE;
+	bool	have_v6 = FALSE;
+	char	address_v4[DNS_SD_ADDRESS_STR_MAX+1] = "";
+	char	address_v6[DNS_SD_ADDRESS_STR_MAX+1] = "";
 	for (CFIndex i = 0; i < CFArrayGetCount(addrArr); i++) {
 		struct sockaddr_in *sa = (struct sockaddr_in *)
-		CFDataGetBytePtr(CFArrayGetValueAtIndex(addrArr, i));
+			CFDataGetBytePtr(CFArrayGetValueAtIndex(addrArr, i));
 		switch(sa->sin_family) {
 			case AF_INET:
 				if (inet_ntop(sa->sin_family, &sa->sin_addr,
-							  dd->address_v4, sizeof(dd->address_v4))) {
-					dd->have_v4 = TRUE;
+							  address_v4, sizeof(address_v4))) {
+					have_v4 = TRUE;
 				}
 			case AF_INET6:
 				if (inet_ntop(sa->sin_family, &sa->sin_addr,
-							  dd->address_v6, sizeof(dd->address_v6))) {
-					dd->have_v6 = TRUE;
+							  address_v6, sizeof(address_v6))) {
+					have_v6 = TRUE;
 				}
 		}
 	}
 
+	if (!have_v4 && !have_v6) {
+		WARNING("DNS SD: Can't resolve valid address for service %s.\n", name);
+		goto exit;
+	}
+
+	/* Set properties on the last element on the list. */
+	while (dd->next)
+		dd = dd->next;
+
+	dd->port = port;
+	dd->hostname = strdup(hostname);
```

Comment:
i can't seem to find any equivalent `free(dd->hostname)`;
does avahi handle that?

Reply:
fixed.